### PR TITLE
TINY-8297: Quickbar toolbar positioned in the middle of table cell selection if contains more than one table cell

### PR DESCRIPTION
**Pull request by shanmen-tiny**
Related Ticket: TINY-8297

Description of Changes:
* Quickbar toolbar positioned in the middle of table cell selection if contains more than one table cell selection

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/`, `hotfix/` or `spike/`

Review:
* [x] Milestone set
* [x] Docs ticket created (if applicable)

GitHub issues (if applicable):

### CODE REVIEW

File: modules/alloy/src/main/ts/ephox/alloy/positioning/mode/SelectionAnchor.ts [hunk 1]
```diff
@@ -15,14 +15,21 @@ import * as ContentAnchorCommon from './ContentAnchorCommon';
 const descendOnce = (element: SugarElement<Node>, offset: number): Descend.ElementAndOffset<Node> =>
   SugarNode.isText(element) ? Descend.point(element, offset) : Descend.descendOnce(element, offset);
 
-const getAnchorSelection = (win: Window, anchorInfo: SelectionAnchor): Optional<SimRange> => {
+const isSimRange = (detail: SimRange | SugarElement<Node>[]): detail is SimRange =>
+  (detail as SimRange).start !== undefined;
```
Comment (TheSpyder) [inline modules/alloy/src/main/ts/ephox/alloy/positioning/mode/SelectionAnchor.ts:19]: `start` seems a bit too generic to use for determining the shape of the object. I suggest using `foffset` as that's highly unlikely to be used anywhere else.
```suggestion
  (detail as SimRange).foffset !== undefined;
```

File: modules/alloy/src/main/ts/ephox/alloy/positioning/mode/SelectionAnchor.ts [hunk 2]
```diff
@@ -32,20 +39,49 @@ const placement = (component: AlloyComponent, anchorInfo: SelectionAnchor, origi
...
-      const x = SugarElement.fromText(Unicode.zeroWidth);
-      Insert.before(sel.start, x);
-      // Certain things like <p><br/></p> with (p, 0) or <br>) as collapsed selection do not return a client rectangle
-      const rect = WindowSelection.getFirstRect(win, SimSelection.exact(x, 0, x, 1));
-      Remove.remove(x);
-      return rect;
-    });
+    if (isSimRange(sel)) {
+      const optRect = WindowSelection.getBounds(win, SimSelection.exactFromRange(sel)).orThunk(() => {
+        const x = SugarElement.fromText(Unicode.zeroWidth);
```
Comment (TheSpyder) [inline modules/alloy/src/main/ts/ephox/alloy/positioning/mode/SelectionAnchor.ts:44]: Better variable name please

**Review by TheSpyder — APPROVED** [review]

File: modules/alloy/src/main/ts/ephox/alloy/positioning/mode/SelectionAnchor.ts [hunk 3]
```diff
@@ -32,20 +39,41 @@ const placement = (component: AlloyComponent, anchorInfo: SelectionAnchor, origi
...
   });
 
   const targetElement = getAnchorSelection(win, anchorInfo)
-    .bind((sel) => SugarNode.isElement(sel.start) ? Optional.some(sel.start) : Traverse.parentElement(sel.start));
+    .bind((sel) => {
+      if (isSimRange(sel)) {
+        return SugarNode.isElement(sel.start) ? Optional.some(sel.start) : Traverse.parentElement(sel.start);
+      }
+
+      return Optional.some(sel.firstCell);
```
Comment (ltrouton) [inline modules/alloy/src/main/ts/ephox/alloy/positioning/mode/SelectionAnchor.ts:75]: Probably best for consistency

```suggestion
      } else {
            return Optional.some(sel.firstCell);
      }
```

File: modules/alloy/CHANGELOG.md
Comment (ltrouton) [inline modules/alloy/CHANGELOG.md:1]: Should we mention the `SelectionAnchorSpec.getSelection` changing as that the update is an API change?

**Review by ltrouton — APPROVED** [review]

File: modules/alloy/CHANGELOG.md
Comment (shanmen-tiny) [reply modules/alloy/CHANGELOG.md:1]: I have made changes to both Alloy and TinyMCE changelog, and have mentioned the `SelectionAnchorSpec.getSelection` changes made in Alloy, and TinyMCE about the `getCursorAnchor` changes

Comment (ltrouton) [reply modules/alloy/CHANGELOG.md:1]: Thanks. Just left a couple of suggestions on the new entries

**Pull request merged**